README.md
[![Build Status](https://travis-ci.org/fkooman/php-lib-rest-plugin-authentication-form.svg?branch=master)](https://travis-ci.org/fkooman/php-lib-rest-plugin-authentication-form)
[![Scrutinizer Code Quality](https://scrutinizer-ci.com/g/fkooman/php-lib-rest-plugin-authentication-form/badges/quality-score.png?b=master)](https://scrutinizer-ci.com/g/fkooman/php-lib-rest-plugin-authentication-form/?branch=master)

# Introduction
HTML Form Authentication plugin for `fkooman/rest-plugin-authentication`.

# License
Licensed under the Apache License, Version 2.0;

   http://www.apache.org/licenses/LICENSE-2.0


CHANGES.md
# Release History

## 4.1.1 (2016-12-18)
- notice, not error for failed authentication

## 4.1.0 (2016-12-18)
- implement Logger

## 4.0.0 (2016-08-19)
- update `fkooman/rest` and `fkooman/http` dependencies

## 3.0.4 (2016-04-12)
- no longer destroy the session, but just delete all authentication 
  keys

## 3.0.3 (2016-03-25)
- update `fkooman/json`

## 3.0.2 (2016-03-07)
- fix running test on newer versions of `fkooman/tpl` dealing with 
  breaking API

## 3.0.1 (2015-12-20)
- if `login_hint` does not match currently logged in user, do not silently
  continue with the already logged in user (issue #1)

## 3.0.0 (2015-11-20)
- finish ability to show credential error in template
- improved session testing
- major cleanups
- API change for contructing the object and for template integration

## 2.0.0 (2015-11-19)
- major API update for new `fkooman/rest-plugin-authentication`

## 1.0.1 (2015-10-27)
- fix a small bug where logout expects a query parameter for `redirect_to`

## 1.0.0 (2015-10-23)
- initial release


tests/fkooman/Rest/Plugin/Authentication/Form/Test/TestSession.php
<?php

/**
 *  Copyright 2015 François Kooman <fkooman@tuxed.net>.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

namespace fkooman\Rest\Plugin\Authentication\Form\Test;

use fkooman\Http\SessionInterface;

class TestSession implements SessionInterface
{
    /** var @array */
    private $s;

    public function __construct()
    {
        $this->s = [];
    }

    public function set($key, $value)
    {
        $this->s[$key] = $value;
    }

    public function delete($key)
    {
        unset($this->s[$key]);
    }

    public function has($key)
    {
        return array_key_exists($key, $this->s);
    }

    public function get($key)
    {
        if (!$this->has($key)) {
            return;
        }

        return $this->s[$key];
    }

    public function destroy()
    {
        $this->s = [];
    }
}


tests/fkooman/Rest/Plugin/Authentication/Form/FormAuthenticationTest.php
<?php

/**
 *  This program is free software: you can redistribute it and/or modify
 *  it under the terms of the GNU Affero General Public License as
 *  published by the Free Software Foundation, either version 3 of the
 *  License, or (at your option) any later version.
 *
 *  This program is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *  GNU Affero General Public License for more details.
 *
 *  You should have received a copy of the GNU Affero General Public License
 *  along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

namespace fkooman\Rest\Plugin\Authentication\Form;

require_once __DIR__.'/Test/TestTemplateManager.php';
require_once __DIR__.'/Test/TestSession.php';

use fkooman\Http\Request;
use fkooman\Http\SessionInterface;
use fkooman\Rest\Plugin\Authentication\AuthenticationPlugin;
use fkooman\Rest\Plugin\Authentication\Form\Test\TestSession;
use fkooman\Rest\Plugin\Authentication\Form\Test\TestTemplateManager;
use fkooman\Rest\Service;
use PHPUnit_Framework_TestCase;

class FormAuthenticationTest extends PHPUnit_Framework_TestCase
{
    public function testAuth()
    {
        $request = new Request(
            [
                'SERVER_NAME' => 'www.example.org',
                'SERVER_PORT' => 80,
                'QUERY_STRING' => '',
                'REQUEST_URI' => '/',
                'SCRIPT_NAME' => '/index.php',
                'REQUEST_METHOD' => 'GET',
            ]
        );
        $testSession = new TestSession();
        $testSession->set('_auth_form_user_name', 'foo');
        $formAuth = $this->getFormAuth($testSession);
        $this->assertEquals('foo', $formAuth->isAuthenticated($request)->getUserId());
    }

    public function testAuthNonMatchingLoginHint()
    {
        $request = new Request(
            [
                'SERVER_NAME' => 'www.example.org',
                'SERVER_PORT' => 80,
                'QUERY_STRING' => 'login_hint=bar',
                'REQUEST_URI' => '/?login_hint=bar',
                'SCRIPT_NAME' => '/index.php',
                'REQUEST_METHOD' => 'GET',
            ]
        );
        $testSession = new TestSession();
        $testSession->set('_auth_form_user_name', 'foo');
        $formAuth = $this->getFormAuth($testSession);
        $this->assertFalse($formAuth->isAuthenticated($request));
    }

    public function testAuthNotAuthenticated()
    {
        $request = new Request(
            [
                'SERVER_NAME' => 'www.example.org',
                'SERVER_PORT' => 80,
                'QUERY_STRING' => 'login_hint=foo',
                'REQUEST_URI' => '/?login_hint=foo',
                'SCRIPT_NAME' => '/index.php',
                'REQUEST_METHOD' => 'GET',
            ]
        );
        $testSession = new TestSession();
        $formAuth = $this->getFormAuth($testSession);
        $this->assertFalse($formAuth->isAuthenticated($request));
        $response = $formAuth->requestAuthentication($request);
        $this->assertSame(
            [
                'HTTP/1.1 200 OK',
                'Content-Type: text/html;charset=UTF-8',
                'X-Frame-Options: DENY',
                "Content-Security-Policy: default-src 'self'",
                'Content-Length: 107',
                '',
                '{"formAuth":{"login_hint":"foo","_auth_form_invalid_credentials":null,"_auth_form_invalid_user_name":null}}',
            ],
            $response->toArray()
        );
        $this->assertNull($testSession->get('_auth_form_user_name'));
    }

    public function testAuthNotAuthenticatedAfterAttempt()
    {
        $request = new Request(
            [
                'SERVER_NAME' => 'www.example.org',
                'SERVER_PORT' => 80,
                'QUERY_STRING' => 'login_hint=foo',
                'REQUEST_URI' => '/?login_hint=foo',
                'SCRIPT_NAME' => '/index.php',
                'REQUEST_METHOD' => 'GET',
            ]
        );
        $testSession = new TestSession();
        $testSession->set('_auth_form_invalid_credentials', true);
        $testSession->set('_auth_form_invalid_user_name', 'fooz');
        $formAuth = $this->getFormAuth($testSession);
        $this->assertFalse($formAuth->isAuthenticated($request));
        $response = $formAuth->requestAuthentication($request);
        $this->assertSame(
            [
                'HTTP/1.1 200 OK',
                'Content-Type: text/html;charset=UTF-8',
                'X-Frame-Options: DENY',
                "Content-Security-Policy: default-src 'self'",
                'Content-Length: 109',
                '',
                '{"formAuth":{"login_hint":"foo","_auth_form_invalid_credentials":true,"_auth_form_invalid_user_name":"fooz"}}',
            ],
            $response->toArray()
        );
        $this->assertNull($testSession->get('_auth_form_user_name'));
    }

    public function testVerifyCorrectCredentials()
    {
        $request = new Request(
            [
                'SERVER_NAME' => 'www.example.org',
                'SERVER_PORT' => 80,
                'HTTP_ACCEPT' => 'text/html',
                'QUERY_STRING' => '',
                'REQUEST_URI' => '/_auth/form/verify',
                'SCRIPT_NAME' => '/index.php',
                'HTTP_REFERER' => 'http://www.example.org/',
                'PATH_INFO' => '/_auth/form/verify',
                'REQUEST_METHOD' => 'POST',
            ],
            [
                'userName' => 'foo',
                'userPass' => 'bar',
            ]
        );
        $service = new Service();
        $testSession = new TestSession();
        $formAuth = $this->getFormAuth($testSession);
        $ap = new AuthenticationPlugin();
        $ap->register($formAuth, 'form');
        $service->getPluginRegistry()->registerDefaultPlugin($ap);
        $response = $service->run($request);
        $this->assertSame(
            [
                'HTTP/1.1 302 Found',
                'Content-Type: text/html;charset=UTF-8',
                'Location: http://www.example.org/',
                '',
                '',
            ],
            $response->toArray()
        );
        $this->assertSame('foo', $testSession->get('_auth_form_user_name'));
        $this->assertNull($testSession->get('_auth_form_invalid_credentials'));
        $this->assertNull($testSession->get('_auth_form_invalid_user_name'));
    }

    public function testVerifyWrongUser()
    {
        $request = new Request(
            [
                'SERVER_NAME' => 'www.example.org',
                'SERVER_PORT' => 80,
                'QUERY_STRING' => '',
                'HTTP_ACCEPT' => 'text/html',
                'REQUEST_URI' => '/_auth/form/verify',
                'SCRIPT_NAME' => '/index.php',
                'HTTP_REFERER' => 'http://www.example.org/',
                'PATH_INFO' => '/_auth/form/verify',
                'REQUEST_METHOD' => 'POST',
            ],
            [
                'userName' => 'fooz',
                'userPass' => 'bar',
            ]
        );
        $service = new Service();
        $testSession = new TestSession();
        $formAuth = $this->getFormAuth($testSession);
        $ap = new AuthenticationPlugin();
        $ap->register($formAuth, 'form');
        $service->getPluginRegistry()->registerDefaultPlugin($ap);
        $response = $service->run($request);
        $this->assertSame(
            [
                'HTTP/1.1 302 Found',
                'Content-Type: text/html;charset=UTF-8',
                'Location: http://www.example.org/',
                '',
                '',
            ],
            $response->toArray()
        );
        $this->assertTrue($testSession->get('_auth_form_invalid_credentials'));
        $this->assertSame('fooz', $testSession->get('_auth_form_invalid_user_name'));
        $this->assertNull($testSession->get('_auth_form_user_name'));
    }

    public function testVerifyWrongPass()
    {
        $request = new Request(
            [
                'SERVER_NAME' => 'www.example.org',
                'SERVER_PORT' => 80,
                'QUERY_STRING' => '',
                'HTTP_ACCEPT' => 'text/html',
                'REQUEST_URI' => '/_auth/form/verify',
                'SCRIPT_NAME' => '/index.php',
                'HTTP_REFERER' => 'http://www.example.org/',
                'PATH_INFO' => '/_auth/form/verify',
                'REQUEST_METHOD' => 'POST',
            ],
            [
                'userName' => 'foo',
                'userPass' => 'baz',
            ]
        );
        $service = new Service();
        $testSession = new TestSession();
        $formAuth = $this->getFormAuth($testSession);
        $ap = new AuthenticationPlugin();
        $ap->register($formAuth, 'form');
        $service->getPluginRegistry()->registerDefaultPlugin($ap);
        $response = $service->run($request);
        $this->assertSame(
            [
                'HTTP/1.1 302 Found',
                'Content-Type: text/html;charset=UTF-8',
                'Location: http://www.example.org/',
                '',
                '',
            ],
            $response->toArray()
        );
        $this->assertTrue($testSession->get('_auth_form_invalid_credentials'));
    }

    public function testLogout()
    {
        $request = new Request(
            [
                'SERVER_NAME' => 'www.example.org',
                'SERVER_PORT' => 80,
                'QUERY_STRING' => '',
                'HTTP_ACCEPT' => 'text/html',
                'REQUEST_URI' => '/_auth/form/logout',
                'SCRIPT_NAME' => '/index.php',
                'PATH_INFO' => '/_auth/form/logout',
                'REQUEST_METHOD' => 'POST',
                'HTTP_REFERER' => 'http://www.example.org/',
            ]
        );
        $service = new Service();
        $testSession = new TestSession();
        $testSession->set('_auth_form_user_name', 'foo');
        $formAuth = $this->getFormAuth($testSession);
        $ap = new AuthenticationPlugin();
        $ap->register($formAuth, 'form');
        $service->getPluginRegistry()->registerDefaultPlugin($ap);
        $response = $service->run($request);
        $this->assertSame(
            [
                'HTTP/1.1 302 Found',
                'Content-Type: text/html;charset=UTF-8',
                'Location: http://www.example.org/',
                '',
                '',
            ],
            $response->toArray()
        );
        $this->assertNull($testSession->get('_auth_form_user_name'));
    }

    public function testLogoutRedirectTo()
    {
        $request = new Request(
            [
                'SERVER_NAME' => 'www.example.org',
                'SERVER_PORT' => 80,
                'QUERY_STRING' => '',
                'HTTP_ACCEPT' => 'text/html',
                'REQUEST_URI' => '/_auth/form/logout',
                'SCRIPT_NAME' => '/index.php',
                'PATH_INFO' => '/_auth/form/logout',
                'REQUEST_METHOD' => 'POST',
                'HTTP_REFERER' => 'http://www.example.org/',
            ],
            [
                'redirect_to' => 'http://my-domain.org/loggedOut',
            ]
        );
        $service = new Service();
        $testSession = new TestSession();
        $testSession->set('_auth_form_user_name', 'foo');
        $formAuth = $this->getFormAuth($testSession);
        $ap = new AuthenticationPlugin();
        $ap->register($formAuth, 'form');
        $service->getPluginRegistry()->registerDefaultPlugin($ap);
        $response = $service->run($request);
        $this->assertSame(
            [
                'HTTP/1.1 302 Found',
                'Content-Type: text/html;charset=UTF-8',
                'Location: http://my-domain.org/loggedOut',
                '',
                '',
            ],
            $response->toArray()
        );
        $this->assertNull($testSession->get('_auth_form_user_name'));
    }

    private function getFormAuth(SessionInterface $session)
    {
        $formAuth = new FormAuthentication(
            function ($userId) {
                // foo:bar
                return 'foo' === $userId ? '$2y$10$L/pq7XXpV54.iAk2EE74deR4yR54yEaZb92.gxH2jDEAZkpcWbhiW' : false;
            },
            new TestTemplateManager(),
            $session
        );

        return $formAuth;
    }
}


tests/fkooman/Rest/Plugin/Authentication/Form/Test/TestTemplateManager.php
<?php

/**
 *  Copyright 2015 François Kooman <fkooman@tuxed.net>.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

namespace fkooman\Rest\Plugin\Authentication\Form\Test;

use fkooman\Json\Json;
use fkooman\Tpl\TemplateManagerInterface;

class TestTemplateManager implements TemplateManagerInterface
{
    public function setDefault(array $templateVariables)
    {
        // NOP
    }

    public function addDefault(array $templateVariables)
    {
        // NOP
    }

    public function render($templateName, array $templateVariables = [])
    {
        return Json::encode(
            [
                $templateName => $templateVariables,
            ]
        );
    }
}


example/form.php
<?php

/**
 * Copyright 2015 François Kooman <fkooman@tuxed.net>.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
require_once dirname(__DIR__).'/vendor/autoload.php';
require_once __DIR__.'/SimpleTpl.php';

use fkooman\Rest\Service;
use fkooman\Rest\Plugin\Authentication\Form\FormAuthentication;
use fkooman\Rest\Plugin\Authentication\Form\FormUserInfo;

// this is a very simple TemplateManagerInterface implementation just returning
// strings, in a real application you would use `fkooman/tpl-twig` for example
$tpl = new SimpleTpl();

// initialize the FormAuthentication class, provide it a function that 
// returns a hash for the provided userName, here it returns the hashed value
// of bar (using https://secure.php.net/password_hash)
$auth = new FormAuthentication(
    function ($userName) {
        return '$2y$10$L/pq7XXpV54.iAk2EE74deR4yR54yEaZb92.gxH2jDEAZkpcWbhiW'; // bar
    },
    $tpl, // the template engine
    array('realm' => 'Example') // the realm for a 401
);

$service = new Service();
$service->getPluginRegistry()->registerDefaultPlugin($auth);

// disable the plugin for the indexPage, we do not need to be logged in 
// there...
$service->get(
    '/',
    function () use ($tpl) {
        return $tpl->render(
            'indexPage',
            array()
        );
    },
    array(
        'fkooman\Rest\Plugin\Authentication\Form\FormAuthentication' => array(
            'enabled' => false,
        ),
    )
);

// here we do need to login to be able to watch this page, the authentication
// plugin takes care of the rest...
$service->get(
    '/welcomePage',
    function (FormUserInfo $u) use ($tpl) {
        return $tpl->render(
            'welcomePage',
            array(
                'user_id' => $u->getUserId(),
            )
        );
    }
);
$service->run()->send();


example/SimpleTpl.php
<?php

/**
 * Copyright 2015 François Kooman <fkooman@tuxed.net>.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
use fkooman\Tpl\TemplateManagerInterface;

class SimpleTpl implements TemplateManagerInterface
{
    public function setDefault(array $templateVariables)
    {
    }

    public function render($templateName, array $templateVariables)
    {
        if ('indexPage' === $templateName) {
            return '<!DOCTYPE html>
                <html lang="en">
                    <head>
                        <meta charset="utf-8">
                        <title>Index</title>
                        <meta name="viewport" content="width=device-width, initial-scale=1.0">
                    </head>
                    <body>
                            <h2>Index</h2>
                            <p>Visit the <a href="welcomePage">Welcome Page</a>, you need to be logged in.</p>
                    </body>
                </html>';
        };

        if ('formAuth' === $templateName) {
            return sprintf('<!DOCTYPE html>
                <html lang="en">
                    <head>
                        <meta charset="utf-8">
                        <title>Sign In</title>
                        <meta name="viewport" content="width=device-width, initial-scale=1.0">
                    </head>
                    <body>
                            <h2>Sign In</h2>
                            <p>Please sign in with your username and password.</p>
                            <form method="post" action="_auth/form/verify">
                                <input type="text" name="userName" placeholder="Username" value="%s" autofocus required>
                                <input type="password" name="userPass" placeholder="Password" required>
                                <button type="submit">Sign in</button>
                            </form>
                    </body>
                </html>',
                $templateVariables['login_hint']
            );
        }

        if ('welcomePage' === $templateName) {
            return sprintf('<!DOCTYPE html>
                <html lang="en">
                    <head>
                        <meta charset="utf-8">
                        <title>Welcome</title>
                        <meta name="viewport" content="width=device-width, initial-scale=1.0">
                    </head>
                    <body>
                            <h2>Welcome</h2>
                            <p>Hello %s!</p>
                            <form method="post" action="_auth/form/logout?redirect_to=/">
                                <input type="submit" value="Logout">
                            </form>
                    </body>
                </html>',
                $templateVariables['user_id']
            );
        }

        return '';
    }
}


src/fkooman/Rest/Plugin/Authentication/Form/FormUserInfo.php
<?php

/**
 * Copyright 2015 François Kooman <fkooman@tuxed.net>.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

namespace fkooman\Rest\Plugin\Authentication\Form;

use fkooman\Rest\Plugin\Authentication\UserInfoInterface;

class FormUserInfo implements UserInfoInterface
{
    /** @var string */
    private $userId;

    public function __construct($userId)
    {
        $this->userId = $userId;
    }

    public function getUserId()
    {
        return $this->userId;
    }
}


src/fkooman/Rest/Plugin/Authentication/Form/FormAuthentication.php
<?php

/**
 * Copyright 2015 François Kooman <fkooman@tuxed.net>.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

namespace fkooman\Rest\Plugin\Authentication\Form;

use fkooman\Http\Exception\BadRequestException;
use fkooman\Http\RedirectResponse;
use fkooman\Http\Request;
use fkooman\Http\Response;
use fkooman\Http\Session;
use fkooman\Http\SessionInterface;
use fkooman\Rest\Plugin\Authentication\AuthenticationPluginInterface;
use fkooman\Rest\Service;
use fkooman\Tpl\TemplateManagerInterface;
use Psr\Log\LoggerInterface;
use Psr\Log\NullLogger;

class FormAuthentication implements AuthenticationPluginInterface
{
    /** @var callable */
    private $retrieveHash;

    /** @var \fkooman\Tpl\TemplateManagerInterface */
    private $templateManager;

    /** @var \fkooman\Http\SessionInterface */
    private $session;

    /** @var \Psr\LoggerInterface */
    private $logger;

    public function __construct(callable $retrieveHash, TemplateManagerInterface $templateManager, SessionInterface $session = null, LoggerInterface $logger = null)
    {
        $this->retrieveHash = $retrieveHash;
        $this->templateManager = $templateManager;
        if (is_null($session)) {
            $session = new Session('form');
        }
        $this->session = $session;
        if (is_null($logger)) {
            $logger = new NullLogger();
        }
        $this->logger = $logger;
    }

    public function isAuthenticated(Request $request)
    {
        $authFormUserName = $this->session->get('_auth_form_user_name');
        if (is_null($authFormUserName)) {
            return false;
        }

        // enforce the login_hint
        $loginHint = $request->getUrl()->getQueryParameter('login_hint');
        if (null !== $loginHint) {
            if ($authFormUserName !== $loginHint) {
                return false;
            }
        }

        return new FormUserInfo($authFormUserName);
    }

    public function init(Service $service)
    {
        $service->post(
            '/_auth/form/verify',
            function (Request $request) {
                $httpReferrer = $request->getHeader('Referer');
                if (null === $httpReferrer) {
                    throw new BadRequestException('Referrer header not sent');
                }

                // delete possibly stale auth session
                $this->session->delete('_auth_form_user_name');
                $this->session->delete('_auth_form_invalid_credentials');
                $this->session->delete('_auth_form_invalid_user_name');

                // validate password
                $userName = $request->getPostParameter('userName');
                $userPass = $request->getPostParameter('userPass');

                $passHash = call_user_func($this->retrieveHash, $userName);
                if (false === $passHash || !password_verify($userPass, $passHash)) {
                    $this->logger->notice(sprintf('invalid credentials for user "%s"', $userName));
                    $this->session->set('_auth_form_invalid_credentials', true);
                    $this->session->set('_auth_form_invalid_user_name', $userName);
                } else {
                    $this->session->set('_auth_form_user_name', $userName);
                    $this->session->delete('_auth_form_invalid_credentials');
                    $this->session->delete('_auth_form_invalid_user_name');
                }

                return new RedirectResponse($httpReferrer, 302);
            },
            [
                'fkooman\Rest\Plugin\Authentication\AuthenticationPlugin' => ['enabled' => false],
            ]
        );

        $service->post(
            '/_auth/form/logout',
            function (Request $request) {
                // delete all authentication information
                $this->session->delete('_auth_form_user_name');
                $this->session->delete('_auth_form_invalid_credentials');
                $this->session->delete('_auth_form_invalid_user_name');

                $rootUrl = $request->getUrl()->getRootUrl();
                $redirectToParameter = $request->getPostParameter('redirect_to');
                $redirectTo = self::validateRedirectTo($rootUrl, $redirectToParameter);

                return new RedirectResponse($redirectTo, 302);
            },
            [
                'fkooman\Rest\Plugin\Authentication\AuthenticationPlugin' => ['enabled' => false],
            ]
        );
    }

    public function requestAuthentication(Request $request)
    {
        $response = new Response(200);
        $response->setHeader('X-Frame-Options', 'DENY');
        $response->setHeader('Content-Security-Policy', "default-src 'self'");
        $response->setBody(
            $this->templateManager->render(
                'formAuth',
                [
                    'login_hint' => $request->getUrl()->getQueryParameter('login_hint'),
                    '_auth_form_invalid_credentials' => $this->session->get('_auth_form_invalid_credentials'),
                    '_auth_form_invalid_user_name' => $this->session->get('_auth_form_invalid_user_name'),
                ]
            )
        );

        return $response;
    }

    private static function validateRedirectTo($rootUrl, $redirectToParameter)
    {
        if (is_null($redirectToParameter)) {
            return $rootUrl;
        }

        // MUST be valid absolute URL
        if (false === filter_var($redirectToParameter, FILTER_VALIDATE_URL)) {
            throw new BadRequestException('invalid redirect_to URL');
        }

        return $redirectToParameter;
    }
}
